### ISSUE TITLE
Improve Logging Capabilities

### ISSUE THREAD

**Issue by MrTyton** [coderwilson/FFX_TAS_Python#188]
Currently, console output has a ton of print statements that are outputting a ton of different information that can be hard to parse

Idea is to create a new logging function that will allow for a INFO debug for people running the TAS, as well as amore VERBOSE that will help for actually debugging.

Issue is that the current verbosity is there because of the lack of unit tests or reproducibility of situations, as often someone will run it, report something is wrong, and then the only way to debug is to actually have the verbose console output on screen at the same time so that we can see what it was thinking.

Either way, having a uniform format for outputting information (rather than what currently happens, which is a rather freeform print statement) would be useful.

If we want to create a INFO and VERBOSE, then there are a few things that should be required

1. VERBOSE should always have an output somewhere, like a log file, so that things can be debugged
2. There needs to be some way of tying the INFO log that's being output on screen to the VERBOSE log that's being output in the log file, so that it's _easy_ for someone debugging to tie the video to the log file (Unique identifier maybe?)
3. Should be easy for programmers to specify what level things are at, and to ensure that even if they're called at the same place that they'll have the same information

What level the program is running at should also be something determined at runtime.

Any other thoughts?

**Comment by orkaboy** [comment]
Any preferences regarding unique identifier vs timestamp for log messages?

As discussed on Discord, it might be good to at least have support for a few different log levels rather than just two (if nothing else, it can be used for tagging different pieces of information differently).

My suggestion in regards to the INFO and VERBOSE syncing is to do this:
- Display INFO, and VERBOSE if specified with a flag in config
- Always write both INFO and VERBOSE to the log file (so the messages will be interleaved). With some kind of timestamp/id it should be easy to bind visual info to the log file.

**Comment by MrTyton** [comment]
I think that having a timestamp is probably better, but it should be synced not to system clock but rather real time in the run. I know that the start time is currently being used for calculations for end of run, which should sync up to real time in the livesplit (not game time), so that would probably give a good visual indiciation (also easier to state happend X hours into run, etc)

WDYT about having a unique identifier per LOG call, to make stack traces easier - eg every log call in specfific function is always tagged w/ same unique ID

How many levels do you think would be useful?

**Comment by MrTyton** [comment]
I think that it's gonna be a good idea to actually have our own clasas so that we can add in whatever formats/etc we want, and have that class call the actual log functions - unsure if the pythong logging does that for us. Does it?

**Comment by orkaboy** [comment]
What do you think of this format? I wrote some code that prints this:

```
[00:00:08.136] NewGame      INFO     ====================================
[00:00:08.137] NewGame      INFO     Starting in
[00:00:08.137] NewGame      INFO     3
[00:00:08.636] NewGame      INFO     2
[00:00:09.137] NewGame      INFO     1
[00:00:09.637] NewGame      INFO     GO!!! Good fortune!
[00:00:09.637] NewGame      INFO     ====================================
[00:00:09.638] NewGame      INFO     Set seed: 183
[00:00:00.000] main         INFO     Timer starts now.
[00:00:03.474] DreamZan     INFO     ### CSR check: True
[00:00:04.543] DreamZan     DEBUG    Checkpoint reached: 1
[00:00:05.176] DreamZan     DEBUG    Checkpoint reached: 2
[00:00:05.710] DreamZan     DEBUG    Checkpoint reached: 3
[00:00:06.711] DreamZan     DEBUG    Checkpoint reached: 4
[00:00:07.278] DreamZan     DEBUG    Checkpoint reached: 5
[00:00:07.514] DreamZan     INFO     Ready to name Tidus
```
Note the time format with hour:minute:second.millisecond. Note that the reference time is reset at the "start of game" tick, so it will be accurate to run-time rather than start of the exe.

**Comment by jlward** [comment]
I've not looked at the code for this change yet, but I really like the time stamp matching the timer. That's really slick

**Comment by orkaboy** [comment]
It's somewhat wonky, but it works. Being able to make custom formatters allow for some nice customization.

**Comment by orkaboy** [comment]
> How many levels do you think would be useful?

The suggested usage is:

- DEBUG: Low level clutter
- INFO: Regular operation of program
- WARNING/ERROR/CRITICAL: Something is going wrong

It's possible to define custom log levels as well (they are really just enum numbers), though I'm not sure it will be needed. Possibly if we want multiple levels of debug info granularity.

**Comment by MrTyton** [comment]
I dig it. Thanks.

Possible to add in line number to the DEBUG output?

**Comment by orkaboy** [comment]
> Possible to add in line number to the DEBUG output?

Yes, should be here: https://docs.python.org/3/library/logging.html#logrecord-attributes

The `%(lineno)d` tag should do just that. It would apply to all logs tho, not just debug.

**Comment by MrTyton** [comment]
It is possible w/ a bit of finagling

See https://stackoverflow.com/questions/14844970/modifying-logging-message-format-based-on-message-logging-level-in-python3

**Comment by MrTyton** [comment]
This can be a followup though, I think it's lower priority.

**Comment by orkaboy** [comment]
Barring future feature requests or changes, I think this one can be closed now?
